File: 2023/Leetcode/algorithms/easy/0541_reverse_string_ii/0541_reverse_string_ii.cs
using System;

public class Solution {
    public string ReverseStr(string s, int k) {
        char[] sList = s.ToCharArray(); // Convert the string into a char array
        
        for (int i = 0; i < sList.Length; i += 2 * k) {
            int end = i + k;
            if (end > sList.Length) {
                end = sList.Length;
            }
            
            // Reverse the first k characters within each 2k block
            ReverseSubstring(sList, i, end);
        }
        
        // Convert the modified array back to a string
        return new string(sList);
    }
    
    private void ReverseSubstring(char[] s, int start, int end) {
        while (start < end) {
            char temp = s[start];
            s[start] = s[end - 1];
            s[end - 1] = temp;
            start++;
            end--;
        }
    }
}

class Program {
    static void Main() {
        Solution solution = new Solution();
        
        Console.WriteLine(solution.ReverseStr("abcdefg", 2)); // Output: "bacdfeg"
        Console.WriteLine(solution.ReverseStr("abcd", 2));    // Output: "bacd"
    }
}

File: 2023/Leetcode/algorithms/easy/0541_reverse_string_ii/0541_reverse_string_ii.py
"""
Created Date: 2023/12/01
Author: @1chooo (Hugo ChunHo Lin)
Status: Accepted
"""

class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        # Convert the string into a list to modify characters
        s_list = list(s)
        
        # Iterate through the string with steps of 2k
        for i in range(0, len(s), 2*k):
            # Reverse the first k characters within each 2k block
            s_list[i:i+k] = reversed(s_list[i:i+k])
            print(''.join(s_list))
        
        # Convert the modified list back to a string
        return ''.join(s_list)

if __name__ == '__main__':
    solution = Solution()
    print(solution.reverseStr("abcdefg", 8))
